signac/contrib/project.py

```diff
@@ -789,9 +785,9 @@ def _contains_job_id(self, job_id):
             True if the job id is initialized for this project.
 
         """
-        # We can rely on the project workspace to be well-formed, so just use
-        # str.join with os.sep instead of os.path.join for speed.
-        return os.path.exists(os.sep.join((self.workspace(), job_id)))
+        # We can rely on the project workspace to be well-formed, so use
+        # os.sep.join instead of os.path.join for performance.
+        return os.path.exists(os.sep.join((self.workspace, job_id)))
```

Comment:
Do we mean to use `os.path.isdir` instead of `os.path.exists` here? That is, files shouldn't be allowed.

If we used `isdir`, symbolic links are followed (as we desire). It looks like this would be the same number of syscalls from the implementation I see:

https://github.com/python/cpython/blob/3751b6b030b4a3b88959b4f3c4ef2e58d325e497/Lib/genericpath.py#L39-L45

https://github.com/python/cpython/blob/3751b6b030b4a3b88959b4f3c4ef2e58d325e497/Lib/genericpath.py#L16-L22

We'd need to investigate the performance impact of this proposed change a little more closely but it might eliminate a possibility of false positives from files while being as efficient as possible.

Reply:
Technically I think you are correct, although I don't think it matters in practice. If a file has been created with that hash we either have a bug in signac (which we probably don't know how to recover from without manual intervention) or a user has been manually monkeying with the workspace, which IMO is not something we need to protect against. We can certainly test out the performance implications, although `isdir` is slightly more complicated since it has to check whether the target is a file or a directory (the `stat.S_ISDIR(st.st_mode)` check) so it will definitely be slower (but hopefully not by much, maybe on the order of 50-100 ns).

Reply:
https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/os_path_isdir.py

This microbenchmark suggests that `os.path.exists` takes about 749 ns, while `os.path.isdir` takes about 815 ns. An interesting result, but nothing actionable here imo. I'll resolve this.